### DISCUSSION TITLE
Best practices for host-to-rootless connections, where sudo/suid is possible?

### DISCUSSION THREAD

**Discussion by adelton** [containers/podman#24823] [General]
Hello,

for development, testing, and debugging, I prefer to use rootless containers but I also from time to time need a way to connect to those rootless containers from the host, as an unprivileged user. I know that I can use
```
$ podman unshare --rootless-netns curl http://10.88.42.42/
```
and the connection will work, seen as coming from 10.88.0.1. But the `podman unshare` is not often practical, for example when I need to check something with a browser or other more heavyweight mechanism.

I know that as an unprivileged user, my processes are not able to manipulate network interfaces on the host. I'm OK with having a sudo or suid helper script to create the ingress to the rootless network, on temporary basis.

I first tried to create a new veth link from the host netns to the rootless netns where the `podman0` interface with 10.88.0.1 lives (for a typical bridge network and pasta setup) but I couldn't get pasta do the proper routing to the individual containers.

So I then decided to create a helper container (actually a pod) and move its end of the veth link that pasta creates to the host netns. That way pastra treats connections from that interface as if they were coming from one of the rootless containers, not knowing that they in fact originate from processes in the host netns.

My current WIP investigation and setup is below. I wonder if this is a reasonable approach to use or if I'm missing some simpler way to create the ingress interface / route on the host?

All commands below are run as unprivileged user; `sudo` commands indicate where we need root privileges.

### Make sure NetworkManager does not touch our interface on the host

A one-off setup to prevent NetworkManager from attempting to handle IP address for the `podman-ingress` interface which we will create.

```
( echo '[keyfile]' ; echo unmanaged-devices=interface-name:podman-ingress ) | sudo tee /etc/NetworkManager/conf.d/podman-ingress.conf
sudo systemctl reload NetworkManager
```

### Create a pod from which we will "steal" the link
```
podman pod create --name podman-ingress --infra-name podman-ingress-infra --network bridge
podman pod start podman-ingress
```

### Get its pid and IP address
```
set -e
INGRESS_INFRA_PID=$( podman inspect podman-ingress-infra | jq -r '.[0].PidFile' )
test -n "$INGRESS_INFRA_PID"
INGRESS_INFRA_PID=$( cat $INGRESS_INFRA_PID )
test -n "$INGRESS_INFRA_PID"
INGRESS_INFRA_IP=$( podman inspect podman-ingress-infra | jq -r '.[0].NetworkSettings | .IPAddress + "/" + (.IPPrefixLen | tostring)' )
test -n "$INGRESS_INFRA_IP"
echo "We will use link from container (pid $INGRESS_INFRA_PID) with $INGRESS_INFRA_IP"
```

### Move the link to the host netns
```
sudo ip netns attach podman-ingress-pod $INGRESS_INFRA_PID
sudo ip -n podman-ingress-pod link set eth0 name podman-ingress netns 1 up
sudo ip addr add $INGRESS_INFRA_IP dev podman-ingress
```

### Check the setup
```
ip a show dev podman-ingress
```

In a different terminal, run a testing container:
```
podman run --rm --network podman --ip 10.88.42.42 docker.io/library/nginx
```

In the first terminal, connect to it:
```
curl http://10.88.42.42/
```

### Tear down the setup

Move the link back to the pod, so that podman does not complain when removing it; avoiding
`network: netavark: failed to delete container veth eth0: Netlink error: No such device (os error 19)`:

```
sudo ip link set podman-ingress netns podman-ingress-pod name eth0
```

Remove the `podman-ingress-pod` netns name:
```
sudo ip netns del podman-ingress-pod
```

Remove the ingress pod:
```
podman pod rm -f podman-ingress
```

**Comment by sbrivio-rh** [comment]
> I also from time to time need a way to connect to those rootless containers from the host

One thing I'm completely missing here is: what prevents you from using plain port forwarding (Podman's `-p` or pasta's `-t`)? If you do that and connect to a loopback address on the host, you'll also reach the container.

**Reply by adelton** [reply]
I typically have multiple containers to work with, all of them using the same (443) port. And they do redirects or connections among themselves -- think SAML or OpenID Connect, or FreeIPA master and replica and clients. When debugging, it is not practical to have an extra layer (port number mapping) to care about.

**Reply by sbrivio-rh** [reply]
So you differentiate between them using different addresses? By the way, you can also map port 443 binding to different addresses to different containers, and with `--freebind` you can bind to arbitrary addresses (even if not locally assigned).

Example with stand-alone pasta(1) for simplicity:

```
# ip route add local 1.2.3.0/24 dev lo # "AnyIP" route
```

```
$ pasta --freebind -t 1.2.3.4/443 --config-net
# nc -l 443
```

```
$ pasta --freebind -t 1.2.3.5/443 --config-net
# nc -l 443
abcd
```

```
$ echo "abcd" | nc 1.2.3.5 443
```

**Reply by adelton** [reply]
> So you differentiate between them using different addresses?

Yes. I basically treat those containers as VMs (and often they actually run systemd and multiple services).

> `pasta --freebind -t 1.2.3.4/443 --config-net`

I'm getting
```
Failed to bind port 443 (Permission denied) for option '-t 1.2.3.4/443', exiting
```

I can use
```
sudo sysctl net.ipv4.ip_unprivileged_port_start=443
```
to change that limit but with that we are already getting to the `sudo` territory. :-)

> `# nc -l 443`

I'm getting
```
Ncat: Failed to resolve default IPv4 address: Name or service not known. QUITTING.
```

Not sure what should or shouldn't have gotten setup by pasta at this point. Running `ip a` in that pasta process shows
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host proto kernel_lo 
       valid_lft forever preferred_lft forever
2: wlp0s20f3: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default qlen 1000
    link/ether 1a:bc:91:ee:40:62 brd ff:ff:ff:ff:ff:ff
```

It feels like playing with individual ports is more work than the wholesale ingress to the rootless network that I'm able to do by moving that veth link to the host netns.

**Reply by sbrivio-rh** [reply]
> I can use
> 
> ```
> sudo sysctl net.ipv4.ip_unprivileged_port_start=443
> ```
> 
> to change that limit but with that we are already getting to the `sudo` territory. :-)

Yes, but that's the recommended way (for the reason explained in pasta(1), `Binding to low numbered ports (well-known or system ports, up to 1023)`) to bind to low ports, and it won't leave interfaces created as root around, so it's arguably the least privileged way to do things.

> I'm getting
> 
> ```
> Ncat: Failed to resolve default IPv4 address: Name or service not known. QUITTING.
> ```
> 
> Not sure what should or shouldn't have gotten setup by pasta at this point. Running `ip a` in that pasta process shows
> 
> ```
> 1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
>     link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
>     inet 127.0.0.1/8 scope host lo
>        valid_lft forever preferred_lft forever
>     inet6 ::1/128 scope host proto kernel_lo 
>        valid_lft forever preferred_lft forever
> 2: wlp0s20f3: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default qlen 1000
>     link/ether 1a:bc:91:ee:40:62 brd ff:ff:ff:ff:ff:ff
> ```

Did you forget `--config-net` perhaps?

**Reply by adelton** [reply]
> 
> Did you forget `--config-net` perhaps?

Ah, you are right, I did.

**Reply by adelton** [reply]
I confirm that the approach with
```
$ sudo ip route add local 10.88.0.0/16 dev lo
```
and then using either
```
$ podman run --rm --network pasta:--ipv4-only,-a,10.88.0.150,-n,16,-g,10.88.0.1,-t,10.88.0.150/80,-m,1500,--no-ndp,--no-dhcpv6,--no-dhcp docker.io/library/nginx
```
or
```
$ podman run --rm -p 10.88.0.150:80:80 --network pasta:--ipv4-only,-a,10.88.0.150,-n,16,-g,10.88.0.1,-m,1500,--no-ndp,--no-dhcpv6,--no-dhcp docker.io/library/nginx
```
works -- I'm then able to
```
$ curl http://10.88.0.150/
```
from the host, and from other containers that have network configured in the same way.

Now I wonder, how to combine that with the rootless bridge setup, to take advantage of the automatic assignment of the IP addresses for containers that are clients, where no well-known externally accessible IP address is needed. I tried adding
```
,--netns,/run/user/1000/containers/networks/rootless-netns/rootless-netns
```
to the `--network pasta:...` parameter list to force the interfaces to be created in the bridge network netns but that does not seem to be the way.

**Reply by sbrivio-rh** [reply]
> ```
> $ podman run --rm -p 10.88.0.150:80:80 --network pasta:--ipv4-only,-a,10.88.0.150,-n,16,-g,10.88.0.1,-m,1500,--no-ndp,--no-dhcpv6,--no-dhcp docker.io/library/nginx
> ```

By the way, you don't strictly need all those `--no-...` options I guess.

> Now I wonder, how to combine that with the rootless bridge setup, to take advantage of the automatic assignment of the IP addresses for containers that are clients, where no well-known externally accessible IP address is needed.

If you want to combine it with a rootless bridge setup, you can create a custom network instead, and the network diagram would look like https://github.com/containers/podman/discussions/22943#discussioncomment-9795883 (details about how to configure all the available rootless modes at https://github.com/eriksjolund/podman-networking-docs).

Or you can use IPv6, create an "AnyIP" route for a whole /64 prefix, and do something similar to the configuration outlined in https://bugs.passt.top/show_bug.cgi?id=48 (not "automatically" supported yet, but you can do something like that manually). Say:

```
# ip route add local 2001:db8:b1bb:1d1::/64 dev lo
```

...then let containers configure their own address via SLAAC (tell pasta to advertise your prefix via NDP) and use those automatically-generated addresses instead.

**Comment by abompard** [comment]
Hey folks, I'm reaching this discussion thread, thanks for it. I have basically the same problem except that I have multiple containers that I need to reach over ports 80/443, so I can't publish the ports to the host. I've added the local route with 
```
$ sudo ip route add local 10.89.0.0/16 dev lo
```

But when I try to reach the containers from the host I get:
```
$ podman container inspect tinystage-ipa | jq '.[0].NetworkSettings.Networks.tinystage.IPAddress'
"10.89.1.6"
$ curl 10.89.1.6
curl: (7) Failed to connect to 10.89.1.6 port 80 after 0 ms: Could not connect to server
```

Is there a way to open the port so I can reach it from the host without publishing it?

**Reply by sbrivio-rh** [reply]
Well, no, you *need* to forward ("publish") it, otherwise pasta won't bind the port at all, see also https://github.com/containers/podman/discussions/24823#discussioncomment-11545969.

If you want to bind a port so that only the host can see it, bind it (publish it) to 127.0.0.1/::1.

**Reply by abompard** [reply]
Ah, I see. Then I suppose I should put a proxy in front of the containers to handle the concurrent ports.

Or should I use a different network mode? I looked at the podman-networking-docs link you provided but nothing seems to match. Is there a way to route network packets to the containers on their original exposed port and using their IP address?

I'm very sorry if I'm missing something obvious, thanks for any help.

**Reply by sbrivio-rh** [reply]
> Ah, I see. Then I suppose I should put a proxy in front of the containers to handle the concurrent ports.

Wait, which concurrent ports? Port 443 in 10.89.1.6:443 is a different port than port 443 in 10.89.1.7:443. And both Podman and pasta give you the possibility of publishing (`-p`) a port binding to a specific address. Note that if you bind low (<= 1023) ports as unprivileged user you need to adjust the related sysctl.

> Or should I use a different network mode?

You can, but the basic rootless thing should work. Just assign all the addresses you need to the host... or don't, but give `--freebind` to pasta in that case. And publish ports bound to specific addresses.

**Reply by abompard** [reply]
> Wait, which concurrent ports? Port 443 in 10.89.1.6:443 is a different port than port 443 in 10.89.1.7:443. And both Podman and pasta give you the possibility of publishing (`-p`) a port binding to a specific address.

Oh, you mean adding IP addresses to the host itself, and then binding the containers' ports to those IPs? Yeah I suppose I could do that, I would need to choose a private ip range and manually assign the containers to IPs in this range. The number of containers I'm going to start isn't fixed, so the manual assignment isn't very practical. But isn't it possible to use the already existing container IPs, in the 10.89. range? The containers can talk to each other on the original ports (443) without port mapping.

Maybe I haven't explained properly what I want to do, indeed I should have.
I currently have a set of VMs running in a bridged network that libvirt maintains, with a private IP range in 192.168. and dnsmasq running. I can access these VMs on any port using their IP (or actually a hostname that vagrant-hostmanager maintains in /etc/hosts). I'd like to convert this setup to containers to save resources and stop using vagrant. I'd really like to keep the original ports because these VMs host complex apps like FreeIPA (that includes kerberos) and an OIDC provider, which implies multiple HTTP redirects on fixed URLs. It's basically a lightweight staging environment.

I thought I could setup a bridged network with podman and access the containers' IPs from the host as I did with libvirt's bridged network, but that's where I'm stuck, I can't find a way to route the queries from the host to the original container ports. I don't need those ports to be published on the host, and I see that containers can connect to each other's ports without publishing them. I'd like for the host to do the same. I hoped adding the route would suffice, but nope.

Am I trying to bend containers to do something they aren't supposed to do?
Thanks for any help, I'm really unfamiliar with all this.

**Reply by sbrivio-rh** [reply]
> > Wait, which concurrent ports? Port 443 in 10.89.1.6:443 is a different port than port 443 in 10.89.1.7:443. And both Podman and pasta give you the possibility of publishing (`-p`) a port binding to a specific address.
> 
> Oh, you mean adding IP addresses to the host itself

No, you can skip that part with the local "AnyIP" route and giving `--freebind` to pasta. At that point, even if the addresses are not configured on the host, the containers can use them.

> and then binding the containers' ports to those IPs? Yeah I suppose I could do that, I would need to choose a private ip range and manually assign the containers to IPs in this range. The number of containers I'm going to start isn't fixed, so the manual assignment isn't very practical. But isn't it possible to use the already existing container IPs, in the 10.89. range? The containers can talk to each other on the original ports (443) without port mapping.

Yes, you can do that.

It's usually more practical with IPv6 (that's what the scenario at https://bugs.passt.top/show_bug.cgi?id=48 describes), but I think it should work with IPv4 as well.

> Maybe I haven't explained properly what I want to do, indeed I should have. I currently have a set of VMs running in a bridged network that libvirt maintains, with a private IP range in 192.168. and dnsmasq running. I can access these VMs on any port using their IP (or actually a hostname that vagrant-hostmanager maintains in /etc/hosts). I'd like to convert this setup to containers to save resources and stop using vagrant. I'd really like to keep the original ports because these VMs host complex apps like FreeIPA (that includes kerberos) and an OIDC provider, which implies multiple HTTP redirects on fixed URLs. It's basically a lightweight staging environment.
> 
> I thought I could setup a bridged network with podman and access the containers' IPs from the host as I did with libvirt's bridged network, but that's where I'm stuck, I can't find a way to route the queries from the host to the original container ports. I don't need those ports to be published on the host, and I see that containers can connect to each other's ports without publishing them. I'd like for the host to do the same. I hoped adding the route would suffice, but nope.

Wait, how do the containers reach each other if you do **not** publish the ports? They are in the same pod/custom network? In that case, it's a bit more complicated, and it looks like this: https://github.com/containers/podman/discussions/22943#discussioncomment-9795883.

> Am I trying to bend containers to do something they aren't supposed to do?

No, it should work, it's a relatively common setup I would say.

**Reply by abompard** [reply]
> No, you can skip that part with the local "AnyIP" route and giving `--freebind` to pasta. At that point, even if the addresses are not configured on the host, the containers can use them.
> It's usually more practical with IPv6 (that's what the scenario at https://bugs.passt.top/show_bug.cgi?id=48 describes), but I think it should work with IPv4 as well.

Yeah, it's 2025 and I still don't know how do handle IPv6 properly :-(
I've tried to read up on AnyIP. Is there a reserved address space for it? Should I use dnsmasq for IPv6 address allocation and DNS in the containers subnet?
If using IPv6 makes things easier in the end I'm happy to use it.

> Wait, how do the containers reach each other if you do **not** publish the ports? They are in the same pod/custom network? In that case, it's a bit more complicated, and it looks like this: [#22943 (comment)](https://github.com/containers/podman/discussions/22943#discussioncomment-9795883).

Oh gosh I haven't mentioned that, sorry! Yeah I have made a custom network for this staging platform, with a simple `podman network create`, because I thought it would be less intrusive with other possible containers on the user's laptop. That said if it complicates things I'm happy to use the default network.

> > Am I trying to bend containers to do something they aren't supposed to do?

> No, it should work, it's a relatively common setup I would say.

That's reassuring, thanks. If I can help document a "nominal" or "standard" setup for this usecase I'm happy to try to do it in the most generic way possible, and write/test some docs.

**Reply by abompard** [reply]
OK, I've tried to make this work for a few days, but I never managed to make it work properly. I recognize that my setup may be a little unusual, because I need it to work on other people's computers, that I don't know about or control. Here are my requirements for the container network:

  1. Containers must be able to reach each other
  1. Containers must be able to access the internet
  1. Containers must be accessible via their hostname from the host and from each other (if they are accessible via IP, I can probably find a solution to get the name resolution working)
  1. Containers must not have to expose their services on the host on non-standard ports (the OpenID Connect setup requires hardcoding urls and ports in the provider config)
  1. The setup must not break the host's default route while being set up
  1. On network shutdown, the host's networking should be back to normal
  1. The setup must not assume interface names
  1. ideally rootless, but it's fine if it requires some root commands
  1. ideally, the setup would keep working when the interface holding the default route goes up/down, or when the default route changes interface (a laptop goes from wifi to ethernet and back).
  1. ideally, the container network would be different from the host's local network for isolation
  1. ideally, we would not have to manually assign static IPs to individual containers, to make it easier to add more containers.

Do you think I could get all that working with pasta? I know it's a big ask for a simple ticket, thanks for any help!

**Reply by sbrivio-rh** [reply]
What you describe is pretty interesting actually, I'll get back to you in a bit (~couple of days).

**Reply by abompard** [reply]
Hey @sbrivio-rh ! Did you get a chance to look at this? Thanks! :-)

**Reply by sbrivio-rh** [reply]
Sorry for the delay, I swear that it had almost surfaced from my to-do list, let's give it a bump now.

After pondering and checking things for a while, I have to say that up to point 10, maybe with a partial exception for points 3. and 4., Podman already supports your requirement, that is:

> 1. Containers must be able to reach each other
> 2. Containers must be able to access the internet

This could be done using multiple containers in a custom network, and the current diagram looks like https://github.com/containers/podman/discussions/22943#discussioncomment-9795883

> 3. Containers must be accessible via their hostname from the host and from each other (if they are accessible via IP, I can probably find a solution to get the name resolution working)

This is taken care of by aardvark-dns in that model, except that, from outside/the host, the containers wouldn't be visible with different IP addresses, which is something you need, I guess.

> 4. Containers must not have to expose their services on the host on non-standard ports (the OpenID Connect setup requires hardcoding urls and ports in the provider config)

You can `--publish` ports without any translation. If you need to bind the same port to different external IP addresses, though, you can't currently do that with a custom network. You could do it with pasta and `--freebind` as I mentioned earlier, but not (yet) together with Podman.

> 5. The setup must not break the host's default route while being set up

With rootless networking, we don't (can't) touch host routes anyway, and...

> 6. On network shutdown, the host's networking should be back to normal

this is also covered, as a consequence.

> 7. The setup must not assume interface names

Unless your applications do, Podman, rootlesskit, and pasta are not really looking at them.

> 8. ideally rootless, but it's fine if it requires some root commands

That would be rootless. A "rootful" alternative covering some of these points would be macvlan, though.

> 9. ideally, the setup would keep working when the interface holding the default route goes up/down, or when the default route changes interface (a laptop goes from wifi to ethernet and back).

Solved, I think. With custom networks, that always worked. With a single "stand-alone" container and pasta, we addressed that a while ago.

> 10. ideally, the container network would be different from the host's local network for isolation

This is the case unless you go with `--net=host`, which would be quite far from the setup you need anyway.

For 11., however (and to completely support points 3. and 4. above):

> ideally, we would not have to manually assign static IPs to individual containers, to make it easier to add more containers.

You would need a combination of pasta supporting free-binding to addresses *and* not doing NAT, in the sense described at https://bugs.passt.top/show_bug.cgi?id=48, but at the same time that would need to play nicely with Podman, which need, I think, the kind of improvement described at https://pad.passt.top/p/DynamicForwardUpdates, plus its integration with Podman.

That is, you could bring up a number of containers with "AnyIP" routes and non-local addresses, but Podman would have no idea how to connect them, or how to forward ports to them. We plan to add this, but it will definitely take a while (patches with bits of implementation indeed welcome).

At the moment I'm not sure which readily implemented model would be closest to the requirements you describe (which I think we could use as some kind of checklist while we progress with the implementations I mentioned above). Maybe the custom network is close enough, or maybe you want to start different containers and then resolve their names and possibly redirect traffic between them with external tools, such as a local DNS resolver/proxy and nft.

**Comment by sshambar** [comment]
@adelton - I liked the idea you had initially about "stealing" interfaces from the container, but I was trying to create a setup where I could route into a local rootless "kind" cluster using it's cloud-provider-kind load balancer as a front-end... Using a container interface caused routing problems, so I decided to use a different approach in the podman rootless-netns:

```
# Create a podman network to reserve a subnet
podman network create reserved

# Start an infra container to maintain the podman rootless namespace
podman pod create --name infra --network podman
podman pod start infra

# Name to podman rootless netns (requires sudo)
while read pid; do sudo ip netns attach user-rootless $pid; break; done \
< <(podman unshare --rootless-netns sh -c 'echo $$; sleep 2')

# Create veth pair in rootless:
podman unshare --rootless-netns ip link add cni0 type veth peer name cni0-peer

# Use sudo to move peer to host ns:
sudo ip -n user-rootless link set cni0-peer name cni0 netns $(ip netns identify $$)

# Get the reserved subnet
subnet=$(podman network inspect reserved | jq -r '.[0].subnets.[0].subnet')

# Generate a pair of ips from $subnet - eg: 10.89.0.0/24 -> ip1=10.89.0.1 ip2=10.89.0.2 
# Assign one to the host link (sudo)
sudo ip addr add 10.89.0.1/24 dev cni0

# and the other to rootless link
podman unshare --rootless-netns ip addr add 10.89.0.2/24 dev cni0

# Bring the links up
sudo ip link set cni0 up
podman unshare --rootless-netns ip link set cni0 up

# Add a default route (higher metric than podman's) to rootless
podman unshare --rootless-netns ip route add default via 10.89.0.1 dev cni0

# Now for any additional podman networks (eg kind), you can create routes to the rootless
subnet=$(podman network inspect kind | jq -r '.[0].subnets.[0].subnet')

# sudo ip route add $subnet via 10.89.0.2 dev cni0
```

The net result is that any subnet with an added route is directly addressable from the host (and the local subnet if the necessary local routes are advertised!), and all ips/ports are available dynamically (unless blocked by the firewall).  This works for both ipv4 and ipv6...

To tear down:
```
# remove the link in rootless (also removes host link and routes)
podman unshare --rootless-netns ip link del cni0

# remove the infra pod
podman pod stop infra && podman pod rm infra

# delete the reserved network
podman network rm reserved

# remove the netns attachment
sudo ip netns del user-rootless
```

I found the technique so effective that I wrote a script to make the whole exercise effortless to setup/teardown (for my test clusters).  The script handles dual-stack networking, and can add and remove podman networks while leaving the host interface in place (and has cleanup).

You can find the script at https://github.com/sshambar/setup-rootless - it has limited configuration, but doesn't yet handle advanced config like mtu / metrics etc (PRs welcome!)

**Comment by andsens** [comment]
@sshambar I can do you one better. Full connectivity inside an unprivileged container without `CAP_SYS_ADMIN`. No `sysctl` issues, no joining the podman namespace. Straight-up routing from the container root namespace. I haven't bothered with DNS so far, because I didn't need it. Maybe someone else can spend another week figuring out that part?

`/etc/containers/containers.conf`
```
[containers]
volumes = ["/proc:/proc"]
default_sysctls = []
```

Run in container:
```
podman network create test --driver macvlan --subnet 10.45.45.0/24 --ip-range 10.45.45.2-10.45.45.253 --route 172.16.0.0/24,10.45.45.254
sudo bash -ec '
mkdir /dev/net
mknod /dev/net/tun c 10 200
chgrp user /dev/net/tun
chmod g+w /dev/net/tun
vde_switch -group user -mod 660 -d --sock "$PWD"
vde_plug2tap -d --sock "$PWD" podman
ip addr add 172.16.0.1/24 dev podman
ip link set podman up
ip route add 10.45.45.0/24 via 172.16.0.2 dev podman
'
podman unshare --rootless-netns bash -ec '
vde_plug2tap --sock "$PWD" podman &
ip addr add 172.16.0.2/24 dev podman
ip link set podman up
ip link add macvlan0 link eth0 type macvlan mode bridge
ip addr add 10.45.45.254/24 dev macvlan0
ip link set macvlan0 up
wait
' &
# test
podman run --log-level trace --network test -d --rm --hostname echo --name echo docker.io/hashicorp/http-echo -text="Hello World"
wget -qO- http://10.45.45.2:5678
```

The following capabilities are needed
```
SETUID
SETGID
CHOWN
FOWNER
DAC_OVERRIDE
MKNOD
NET_ADMIN
AUDIT_WRITE
```

Other than that you just need https://github.com/virtualsquare/vde-2
